### PR TITLE
[WIP] Define new "ingressVisibility" rule type

### PR DESCRIPTION
**Pull request by rlenglet**
Define a new "ingressVisibility" rule type to forward ingress traffic to
L4 ports through the L7 proxy for access logging.
    
Compute the list of all active ingress visibility rules for each
L4 port, and report them in the endpoint policy API.

```release-note
Introduce ingressVisibility policy rule type to forward ingress traffic to L4 ports at L7 for access logging
```

Signed-off-by: Romain Lenglet <romain@covalent.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/1540)

### CODE REVIEW

**Comment by rlenglet** [comment]
I still need to test this, but I'd like to get early feedback, esp. on the API change.

File: pkg/policy/l4.go [hunk 1]
```diff
@@ -43,8 +48,49 @@ type L4Filter struct {
 	Ingress bool
 }
 
-// CreateL4Filter creates an L4Filter based on an api.PortRule and api.PortProtocol
-func CreateL4Filter(rule api.PortRule, port api.PortProtocol, direction string, protocol string) L4Filter {
+// createHTTPRule creates an AuxRule based on an api.PortRuleHTTP.
```
Comment (aanm) [inline pkg/policy/l4.go]: Can you write an example of what the result would be for a particular `api.PortRuleHTTP`?

File: pkg/policy/api/rule.go [hunk 2]
```diff
@@ -201,9 +201,18 @@ type PortRule struct {
...
 type L7Rules struct {
-	// HTTP specific rules.
+	// Logging rule. Ensures the traffic to the ports is forwarded at L7 for
+	// the purpose of logging requests and responses into the access log.
+	// The value identifies the L7 protocol. Accepted values: "http".
+	// If any non-logging rule is already defined for the ports, all
+	// logging rules for those ports are ignored.
+	//
+	// +optional
+	Logging string `json:"logging,omitempty"`
```
Comment (tgraf) [inline pkg/policy/api/rule.go]: I was considering to suggest to add this as a flag to the individual protocol specific members instead but having this separate structure would allow to support special values such as "auto-detect" to try and guess based on the first 500 bytes or to support specifying a list of protocols to attempt parsing with, e.g. `"http, gprc"`

**Comment by rlenglet** [comment]
[Build 12](https://jenkins.cilium.io/job/cilium/job/cilium/job/PR-1540/12/) hit a deadlock, which seems unrelated to my changes. Investigated with @ianvernon, and found a threading bug that could be the cause. @ianvernon will fix that.

**Comment by rlenglet** [comment]
@ianvernon is fixing the deadlocks that the tests are hitting in #1858. Those are due to the use of the `-l ...` option in `cilium endoint get`, which is not covered by previous tests.
This PR is ready for review.

File: pkg/policy/api/rule.go [hunk 3]
```diff
@@ -313,9 +347,9 @@ type CIDRRule struct {
 
 // L7Rules is a union of port level rule types. Mixing of different port
 // level rule types is disallowed, so exactly one of the following must be set.
-// If none are specified, then no additional port level rules are applied.
+// If none is specified, then no additional port-level rules are applied.
```
Comment (joestringer) [inline pkg/policy/api/rule.go]: nit: `none` isn't a value, and `are` reads easier.

File: pkg/policy/l4.go [hunk 4]
```diff
@@ -23,27 +23,17 @@ import (
...
-type L7ParserType string
-
-const (
-	// ParserTypeHTTP specifies a HTTP parser type
-	ParserTypeHTTP L7ParserType = "http"
-	// ParserTypeKafka specifies a Kafka parser type
-	ParserTypeKafka L7ParserType = "kafka"
-)
+// L7DataMap contains a map of L7 rules per endpoint where key is a string representation of EndpointSelector
+// TODO: Add the EndpointSelector into each map value.
```
Comment (joestringer) [inline pkg/policy/l4.go]: I assume this is something you intend to do during rebase against the latest Kafka changes?

File: pkg/policy/repository.go [hunk 5]
```diff
@@ -93,8 +97,23 @@ loop:
...
 	}
-
-	state.trace(p, ctx)
+	state.trace(p, &ctx)
+
+	if decision == api.Undecided {
+		ctx.PolicyTrace("\nResolving L4 policy for matching on labels\n")
+		// The calls to rule.canReach above check canReachEntities, FromEntities, etc.
+		// But that's NOT sufficient to determine reachability because resolving the L4 policy
+		// may result in the synthesis of new rules that may allow broader reachability.
```
Comment (joestringer) [inline pkg/policy/repository.go:106]: I think this is partially a problem of naming. `CanReachRLocked` is intended to determine only labels-based L3 reachability (based on how it's used through pkg/policy/endpoint). This is used to populate the `bpf.policymap` with entries such as {securityid,0,0} (in which case L4 is irrelevant), or alternatively from policytrace to determine whether pure labels-based policy allows the traffic, as part of `AllowsRLocked` which already does L4 policy trace after it does the call to this function.

In the former (bpf.policymap l3 labels population) case, if the policy allows traffic between the {From,To} endpoint pairs then the behaviour will be the same. If the policy doesn't allow such traffic, then now we'll do this extra `ResolveL4Policy` walk of the repository, then subsequently call `policy.IngressCoversContext` with `ctx.DPorts == nil`, which will almost immediately terminate with an api.Denied result. This is behaviourally correct, but less efficient than today. I realise that now `ResolveL4Policy` is resolving the l3-only policies as well, which means that we can get the information from there, but as far as I can tell this makes no visible impact on policy application. 

In the latter tracing case we now trace the ingress policy multiple times.

To summarise the above, this seems to increase the processing but I don't understand any benefit. To reap the code structure benefits, then I think we should  move more towards a single policy resolution per Consumable which provides all of the relevant L3, L4, CIDR, fromEntities, etc. policies.

File: pkg/policy/repository.go [hunk 6]
```diff
@@ -226,24 +261,23 @@ func (p *Repository) allowsL4Ingress(ctx *SearchContext) api.Decision {
...
-		l4Ingress := p.allowsL4Ingress(ctx)
+	ctx.PolicyTrace("\nResolving L4 policy for matching on ports\n")
 
-		// Explicit deny should deny; Allow+Undecided should allow
-		if l4Egress == api.Denied || l4Ingress == api.Denied {
-			decision = api.Denied
-		} else if l4Egress == api.Allowed || l4Ingress == api.Allowed {
-			decision = api.Allowed
-		}
+	// TODO: Avoid resolving the L4 policy twice, once in each of the following calls.
```
Comment (joestringer) [inline pkg/policy/repository.go:272]: +1

File: pkg/policy/repository.go [hunk 7]
```diff
@@ -226,24 +261,23 @@ func (p *Repository) allowsL4Ingress(ctx *SearchContext) api.Decision {
...
-	decision := p.CanReachRLocked(ctx)
+	decision := p.CanReachRLocked(*ctx)
 	ctx.PolicyTrace("Label verdict: %s", decision.String())
 	if decision == api.Allowed {
 		return decision
 	}
 
-	// We only report the overall decision as L4 inclusive if a port has
-	// been specified
-	if len(ctx.DPorts) != 0 {
```
Comment (joestringer) [inline pkg/policy/repository.go:237]: Why do we need to resolve L4 policy if we don't have any L4 ports to determine L4 policies that might apply?

File: pkg/policy/l4.go [hunk 8]
```diff
@@ -121,7 +121,12 @@ func CreateL4Filter(fromEndpoints []api.EndpointSelector, rule api.PortRule, por
 	return l4
 }
 
-// IsRedirect returns true if the L4 filter contains a port redirection
+// MatchesAnyPort returns true if the L4 filter matches any port.
+func (l4 *L4Filter) MatchesAnyPort() bool {
```
Comment (tgraf) [inline pkg/policy/l4.go:166]: The name was misleading me because I was expecting the skip to only happen when both port and protocol are any. This is in fact what is happening but the name doesn't account for that.

File: pkg/policy/repository_test.go [hunk 9]
```diff
@@ -579,7 +719,61 @@ Label verdict: denied
...
+* Rule {"matchLabels":{"any:bar":""}}: selected
+    No active ingress visibility rules
+* Rule {"matchLabels":{"any:bar":""}}: selected
+    No active ingress visibility rules
+* Rule {"matchLabels":{"any:bar":""}}: selected
+    No active ingress visibility rules
+3 rules matched
+3/3 rules selected
+Found allow rule
+Label verdict: denied
```
Comment (joestringer) [inline pkg/policy/repository_test.go:743]: Now that L4 policy is mixed into L3 resolution, we get output like this stating that the label verdict is denied, even though this should not have changed with the VisibilityRules changes.

**Review by joestringer — COMMENTED** [review]
There's a bunch of great fixes in here, and the ingress visibility rules feature looks really interesting.

My main concerns with this series as-is are:
1) The l3/label-based policy resolution is triggering rebuild of L4policy for each consumer; and
2) the policy trace output is becoming more complicated.

For (1), part of the problem is that the Repository doesn't have a general way to resolve all policy into the relevant l3/l4 datapath representations for the current context (eg, SecurityID and all potential consumers). Pure label-based policy is done by directly iterating the repository rules, while L4 policy is first resolved then relevant pieces pushed down. This series goes one step in the right direction of making all paths do more complete policy resolution---ie there's less diference between `CanReachRLocked` and `AllowsRLocked`, but I don't think it quite goes far enough because now we have each consumer triggering `AllowsLabelAccess` -> `CanReachRLocked()` -> `ResolveL4Policy`, yet that's the path that explicitly doesn't need L4 information. Instead, the L4 policy resolution in `regenerateConsumable` should be able to filter to all of the relevant policies for this Consumable, then re-use this to build the policy map contents for all of its Consumers. This whole item though seems like a refactoring piece that could (should?) be done separate from the new L7VisibilityRules feature.

For (2), I'm not sure quite what to do. The user specifies a list of rules. Previously, in an attempt to match the datapath logic somewhat, we have been breaking this down into firstly, labels-based L3 matching using `CanReachRLocked()`, then Egress L4 using `allowsL4Egress()`, then Ingress L4 (including l3-dependent l4) using `allowsL4Ingress()`[0]. These three sections, even if verbose, each give a summary afterwards that describes what aspect of the policy was looked at:

```
1/1 rules selected
Found no allow rule
Label verdict: undecided
```

Maybe as we solve (1) above, we can shift more to a single-resolution-of-policy with one giant walk of all of the rules being outputted, printing only the selected rules (and their constraints), then print the "label verdict: ... ; l4 egress verdict: ... ; l4 ingress verdict: ..." bits right near the end. Those summaries would then give a more datapath-centric view of how we determined the final verdict (for example, whether we expect that the traffic would be dropped before being forwarded towards its destination, dropped at the destination, or allowed all the way through.)

[0] Let's overlook the missing bits for now (eg, CIDR trace #1292).

File: pkg/policy/l4.go (hunk 4 above)
Comment (rlenglet) [reply pkg/policy/l4.go]: Correct.

File: pkg/policy/repository.go (hunk 5 above)
Comment (rlenglet) [reply pkg/policy/repository.go:106]: I completely agree with the idea of deriving a single unified "Policy" object that covers everything, incl. L3, L4, CIDR, etc. and do all searches on _that_.
It seemed out of scope for this particular PR, and will need a lot more thought. The good news is that this refactoring can be completely confined inside the Repository & rule implementations.

File: pkg/policy/repository.go (hunk 7 above)
Comment (rlenglet) [reply pkg/policy/repository.go:237]: That's because the resolution of the ingress visibility rules synthesize new L3-only rules in some cases. More specifically, in case the policy enforcement mode is "default", and there are no Ingress or Egress rules matching, but there are one or more ingress visibility rules matching.
In this case, the `L4Policy` will contain one "allow-all" Ingress rule and one "allow-all" Egress rule created by the visibility resolution. Those rules are synthesized, so they can't be evaluated just by iterating on the `Repository.rules`. That requires resolving the `L4Policy`.

That's another argument for designing a new "Policy" type that encompasses all of entities, CIDR, L3, L4, visibility, etc. and that should be used for all decisions and tracing.

File: pkg/policy/repository_test.go (hunk 9 above)
Comment (rlenglet) [reply pkg/policy/repository_test.go:743]: Yeah, I'm not happy with the extra trace output.

File: pkg/policy/api/rule.go [hunk 10]
```diff
@@ -266,10 +282,28 @@ type PortProtocol struct {
...
+	// ToPorts is a list of L4 port/protocol.
+	//
+	// +optional
+	ToPorts []PortProtocol `json:"toPorts,omitempty"`
+
+	// L7Protocol identifies the L7 protocol of the ingress traffic to the
+	// ports. Case insensitive. Accepted values: "http", "kafka".
+	//
+	// +optional
+	L7Protocol L7ParserType `json:"l7protocol,omitempty"`
```
Comment (aanm) [inline pkg/policy/api/rule.go]: `s/l7protocol/l7Protocol`

File: pkg/policy/repository.go [hunk 11]
```diff
@@ -152,6 +153,17 @@ func (p *Repository) ResolveL4Policy(ctx *SearchContext) (*L4Policy, error) {
 		}
 	}
 
+	ctx.PolicyTrace("Resolving ingress visibility rules for %+v\n", ctx.To)
```
Comment (aanm) [inline pkg/policy/repository.go:174]: @joestringer IIRC you've removed the `\n` of the tracing messages, right?

File: pkg/policy/rule.go [hunk 12]
```diff
@@ -81,23 +86,67 @@ func (r *rule) sanitize() error {
...
+		l4.FromEndpoints = nil
+		return false
 	}
 
-	policy.FromEndpoints = append(policy.FromEndpoints, fromEndpoints...)
+	l4.FromEndpoints = append(l4.FromEndpoints, fromEndpoints...)
 	return false
 }
 
+func mergeIngressVisibilityPort(ctx *SearchContext, p api.PortProtocol, l7Parser api.L7ParserType,
```
Comment (aanm) [inline pkg/policy/rule.go:112]: Can you document what are the meaning of the return values?

File: pkg/policy/l4.go [hunk 13]
```diff
@@ -159,10 +160,33 @@ func (l4 L4Filter) matchesLabels(labels labels.LabelArray) bool {
...
-// L4PolicyMap is a list of L4 filters indexable by protocol/port
-// key format: "port/proto"
+// L4PolicyMap is a list of L4 filters indexed by L4 port/protocol.
+// Key format: "port/proto".
 type L4PolicyMap map[string]L4Filter
 
+// L7VisibilityRule is an active ingress visibility rule.
+type L7VisibilityRule struct {
+	// Port is the destination L4 port to redirect to L7 for access logging.
+	Port int
```
Comment (aanm) [inline pkg/policy/l4.go]: Why not `uint16`?

File: pkg/policy/rule.go [hunk 14]
```diff
@@ -383,27 +404,67 @@ func (r *rule) resolveL4Policy(ctx *SearchContext, state *traceState, result *L4
 	return nil, nil
 }
 
+// allowAllRule is a L4 rule that allows all ingress & egress L4 traffic.
+var allowAllRule rule = rule{
```
Comment (aanm) [inline pkg/policy/rule.go]: Can we guarantee that we don't modify this object? The best way would have to create a function that would return a new allowAllRule

```
func AllowAllRule() *rule{
    return &rule{...
}
```

File: pkg/endpoint/policy.go [hunk 15]
```diff
@@ -148,9 +148,26 @@ func getSecurityIdentities(owner Owner, selector *api.EndpointSelector) []policy
...
+	} else {
+		port = uint16(filter.Port)
+		var protoUint8 u8proto.U8proto
+		protoUint8, err = u8proto.ParseProtocol(string(filter.Protocol))
+		proto = uint8(protoUint8)
+	}
+	return
+}
+
 func (e *Endpoint) removeOldFilter(owner Owner, filter *policy.L4Filter) int {
```
Comment (aanm) [inline pkg/endpoint/policy.go:171]: What's this function returning?

File: pkg/policy/l4.go [hunk 16]
```diff
@@ -217,6 +223,12 @@ func (l4 L4PolicyMap) containsAllL3L4(labels labels.LabelArray, ports []*models.
 		return api.Denied
 	}
 
+	// Check any rule that accepts any port.
+	anyPortFilter, match := l4["0/ANY"]
```
Comment (aanm) [inline pkg/policy/l4.go]: maybe we should have `"0/ANY"` as a const?

File: pkg/policy/rule.go [hunk 17]
```diff
@@ -371,28 +421,40 @@ func (r *rule) resolveL4Policy(ctx *SearchContext, state *traceState, result *L4
 	state.selectRule(ctx, r)
 	found := 0
 
-	if !ctx.EgressL4Only {
-		if len(r.Ingress) == 0 {
-			ctx.PolicyTrace("    No L4 rules\n")
+	// Always resolve both ingress and egress, in order to always generate the same
+	// L4Policy. This is required for resolving visibility rules. But disable tracing
+	// for the rules that were not requested for tracing.
+	ingressCtx := *ctx
```
Comment (aanm) [inline pkg/policy/rule.go:430]: Be aware you are *not* making a deep copy of `From` and `To`, can you guarantee that `mergeL4` will never modify the `From` nor `To`?

File: pkg/policy/l4.go [hunk 18]
```diff
@@ -61,13 +61,10 @@ func (dm L7DataMap) addRulesForEndpoints(rules api.L7Rules,
...
 	if len(fromEndpoints) > 0 {
 		for _, ep := range fromEndpoints {
-			hash, err := ep.Hash()
-			if err != nil || hash == 0 {
-				return fmt.Errorf("Could not hash (%d) endpoint %e", hash, err)
-			}
-			dm[hash] = api.L7Rules{
-				HTTP:  append(dm[hash].HTTP, rules.HTTP...),
-				Kafka: append(dm[hash].Kafka, rules.Kafka...),
+			str := ep.String()
```
Comment (aanm) [inline pkg/policy/l4.go]: Why not the `Hash()`?

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/policy/repository.go (hunk 7 above)
Comment (joestringer) [reply pkg/policy/repository.go:237]: OK, thanks for the explanation.

File: pkg/policy/repository.go (hunk 11 above)
Comment (joestringer) [reply pkg/policy/repository.go:174]: A lot of them I did, based on unit testing with the new policy trace examples in `pkg/policy/repository_test.go`. I think we'll need to take another look at the policy trace output after this.

File: pkg/policy/rule.go (hunk 17 above)
Comment (joestringer) [reply pkg/policy/rule.go:430]: I think it'd be nice to add some guarantee of this, since I'd expect that the SearchContext should be purely an input to the resolution to filter down which policies are relevant.

File: pkg/policy/api/rule.go (hunk 3 above)
Comment (rlenglet) [reply pkg/policy/api/rule.go]: Done.

File: pkg/policy/api/rule.go (hunk 10 above)
Comment (rlenglet) [reply pkg/policy/api/rule.go]: Done.

File: pkg/policy/l4.go (hunk 13 above)
Comment (rlenglet) [reply pkg/policy/l4.go]: Done. I was mimicking another parts of the API where `int` is used (RedirectPort).

Comment (rlenglet) [reply pkg/policy/l4.go]: Done.

Comment (rlenglet) [reply pkg/policy/l4.go]: I was mimicking other places where `int` is used, e.g. `L4Filter.Port`.

File: pkg/policy/rule.go (hunk 12 above)
Comment (rlenglet) [reply pkg/policy/rule.go:112]: Done.

File: pkg/policy/rule.go (hunk 14 above)
Comment (rlenglet) [reply pkg/policy/rule.go]: I prefer not to create a rule every time.

File: pkg/policy/rule.go (hunk 17 above)
Comment (rlenglet) [reply pkg/policy/rule.go:430]: We never modify the pointed variables, only fields directly in SearchContext.

**Comment by rlenglet** [comment]
Got a green build: [Build 14](https://jenkins.cilium.io/job/cilium/job/cilium/job/PR-1540/14/). Two more build are underway for minor cosmetic cleanups: [Build 15](https://jenkins.cilium.io/job/cilium/job/cilium/job/PR-1540/15/), [Build 16](https://jenkins.cilium.io/job/cilium/job/cilium/job/PR-1540/16/).

**Comment by rlenglet** [comment]
3 last builds are green.

File: pkg/policy/api/rule.go [hunk 19]
```diff
@@ -293,10 +309,28 @@ type PortProtocol struct {
...
```
Comment (tgraf) [inline pkg/policy/api/rule.go]: What is the effect if I specify `ToPorts` but no `L7Protocol`?

File: pkg/policy/api/rule_validation.go [hunk 20]
```diff
@@ -43,6 +44,12 @@ func (r Rule) Sanitize() error {
 		}
 	}
 
+	for _, v := range r.IngressVisibility {
+		if err := v.sanitize(); err != nil {
```
Comment (tgraf) [inline pkg/policy/api/rule_validation.go:48]: Won't this run `sanitize()` on a copy of the `IngressVisibilityRule` so the following in `sanitize()` will have no effect:
```
v.L7Protocol, err = ParseL7ParserType(string(v.L7Protocol))
```

File: pkg/policy/api/rule.go [hunk 21]
```diff
@@ -293,10 +309,28 @@ type PortProtocol struct {
...
+type IngressVisibilityRule struct {
+	// ToPorts is a list of L4 port/protocol.
+	//
+	// +optional
+	ToPorts []PortProtocol `json:"toPorts,omitempty"`
+
+	// L7Protocol identifies the L7 protocol of the ingress traffic to the
+	// ports. Case insensitive. Accepted values: "http", "kafka".
+	//
+	// +optional
```
Comment (tgraf) [inline pkg/policy/api/rule.go:326]: Is this really optional? Sanitize seems to enforce that this value is either http or kafka.

**Review by tgraf — CHANGES_REQUESTED** [review]
I've looked through the API for now.

The changed policy trace is somewhat of a blocker for me. It's basically at the point where this is unusable for anything that involves more than a couple of trivial policies.

Can add a flag which enables tracing of the visibility part? The main use case for policy trace is to figure out if and why policy is denied or allowed for a particular flow. The visibility rules may matter but are secondary and tracing of such could be enabled via a simple flag.

**Review by raybejjani — APPROVED** [review]
pkg/proxy approval on the name changes

File: pkg/policy/rule.go (hunk 17 above)
Comment (aanm) [reply pkg/policy/rule.go:430]: But this will be a problem in the future, I (or any external developer) might not know that we never modify the pointed variables. IMO it doesn't make sense to only be able to change some parts of a struct but not the others.

Comment (rlenglet) [reply pkg/policy/rule.go:430]: I agree that this ugly. And there are currently several places in the policy package where the SearchContext is modified. I'm planning a redesign of the policy code. Tracing needs to be fixed anyway so we report which _set of rules_ denies or allows a flow, in order to troubleshoot policies. That will require invasive changes.

Comment (rlenglet) [reply pkg/policy/rule.go:430]: I will work on that in the coming release. It's just too big for this particular PR.

**Review by aanm — APPROVED** [review]
